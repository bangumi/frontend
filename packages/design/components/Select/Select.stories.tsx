import type { Meta, StoryObj } from '@storybook/react';
import React from 'react';

import Select from '.';

const componentMeta: Meta<typeof Select> = {
  title: 'Modern/Select',
  component: Select,
};

export default componentMeta;

export const Template: StoryObj<typeof Select> = {
  render: () => (
    <Select
      style={{
        width: 400,
      }}
      options={[
        { label: '名字', value: 'name' },
        { label: '年龄', value: 'age' },
        { label: '学校', value: 'school' },
      ]}
      defaultValue='age'
    />
  ),
};
