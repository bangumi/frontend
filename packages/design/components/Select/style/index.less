@import '../../../theme/base';

.bgm-select {
  display: inline-flex;
  box-sizing: border-box;
  position: relative;
  padding: 0.5rem 1rem;
  border: 2px solid @gray-10;
  border-radius: 12px;
  user-select: none;
  font-size: 1rem;
  vertical-align: bottom;
  cursor: pointer;

  &__dropdown {
    position: absolute;
    margin-top: 0.5rem;
    top: 100%;
    border: 2px solid @gray-10;
    box-sizing: border-box;
    padding: 8px 12px;
    width: 100%;
    left: 0;
    border-radius: 19px;
    background: white;

    > div {
      padding: 9px 0;
    }
  }

  select {
    position: absolute;
    left: 0;
    top: 0;
    height: 100%;
    width: 100%;
    opacity: 0;
  }

  &__box {
    display: flex;
    flex: 1;
    align-items: center;
    justify-content: space-between;

    span {
      font-size: 1rem;
      line-height: 1.5rem;
      color: @gray-100;
    }
  }

  &-arrow {
    display: inline-block;
  }

  .bgm-input-group & {
    &:not(:last-of-type) {
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
      border-right: 0;
    }
  }
}
