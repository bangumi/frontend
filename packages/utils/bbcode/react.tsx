import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
