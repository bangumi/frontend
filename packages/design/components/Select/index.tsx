import './style';

import cn from 'classnames';
import { keyBy } from 'lodash';
import React, { useMemo, useState } from 'react';

import { ArrowDown } from '@bangumi/icons';

interface Option {
  label: string;
  value: string;
}

interface SelectProps {
  defaultValue: string;
  style?: React.CSSProperties;
  options: Option[];
  className?: string;
  onChange?: (value: Option | undefined) => void;
}

const Select = ({ options, style, defaultValue, className, onChange }: SelectProps) => {
  const [open, setOpen] = useState(false);
  const [active, setActive] = useState(defaultValue);
  const handleClick = () => {
    setOpen(!open);
  };

  const optionsMap = useMemo(() => keyBy(options, 'value'), [options]);

  const handleOptionClick = (value: string) => {
    setActive(value);
    onChange?.(optionsMap[value]);
  };

  return (
    <div className={cn('bgm-select', className)} onClick={handleClick} style={style}>
      <div className='bgm-select__box'>
        <span>{optionsMap[active]?.label}</span>
        <ArrowDown className='bgm-select-arrow' />
      </div>
      {open ? (
        <ul className={cn('bgm-select__dropdown')} defaultValue={defaultValue}>
          {options.map((option) => (
            <li
              className={cn('bgm-select__dropdown-item')}
              key={option.value}
              value={option.value}
              onClick={() => {
                handleOptionClick(option.value);
              }}
            >
              {option.label}
            </li>
          ))}
        </ul>
      ) : (
        <></>
      )}
    </div>
  );
};

export default Select;
