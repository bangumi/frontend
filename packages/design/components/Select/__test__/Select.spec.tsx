import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import Select from '..';

describe('Select Component', () => {
  it('render properly', () => {
    const { container } = render(
      <Select
        defaultValue='name'
        options={[
          { label: 'name', value: 'name' },
          { label: 'age', value: 'age' },
        ]}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Open dropdown & Select option', () => {
    const onChange = vi.fn();
    const element = render(
      <Select
        defaultValue='name'
        options={[
          { label: 'name', value: 'name' },
          { label: 'age', value: 'age' },
        ]}
        onChange={onChange}
      />,
    );

    fireEvent.click(element.getByText('name'));
    fireEvent.click(element.getByText('age'));
    expect(onChange).toBeCalledTimes(1);
  });
});
