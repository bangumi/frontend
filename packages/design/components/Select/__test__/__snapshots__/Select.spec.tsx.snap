// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select Component > render properly 1`] = `
<div>
  <div
    class="bgm-select"
  >
    <div
      class="bgm-select__box"
    >
      <span>
        name
      </span>
      <div
        class="bgm-select-arrow"
      />
    </div>
    <select>
      <option
        selected=""
        value="name"
      >
        name
      </option>
      <option
        value="age"
      >
        age
      </option>
    </select>
  </div>
</div>
`;
